src/persist/src/indexed/encoding.rs

```diff
@@ -200,34 +200,108 @@ pub enum BlobTraceUpdates {
+            BlobTraceUpdates::Structured { diffs, .. } => diffs,
+        }
     }
 
     /// Return the [`ColumnarRecords`] of the blob.
-    pub fn records(&self) -> &ColumnarRecords {
+    pub fn records(&self) -> Option<&ColumnarRecords> {
         match self {
-            BlobTraceUpdates::Row(c) => c,
-            BlobTraceUpdates::Both(c, _structured) => c,
+            BlobTraceUpdates::Row(c) => Some(c),
+            BlobTraceUpdates::Both(c, _structured) => Some(c),
+            BlobTraceUpdates::Structured { .. } => None,
         }
     }
 
     /// Return the [`ColumnarRecordsStructuredExt`] of the blob.
     pub fn structured(&self) -> Option<&ColumnarRecordsStructuredExt> {
         match self {
             BlobTraceUpdates::Row(_) => None,
-            BlobTraceUpdates::Both(_, structured) => Some(structured),
+            BlobTraceUpdates::Both(_, s) => Some(s),
+            BlobTraceUpdates::Structured { key_values, .. } => Some(key_values),
         }
     }
 
     /// Return the estimated memory usage of the raw data.
     pub fn goodbytes(&self) -> usize {
-        self.records().goodbytes() + self.structured().map_or(0, |e| e.goodbytes())
+        match self {
+            BlobTraceUpdates::Row(c) => c.goodbytes(),
+            // NB: we only report goodbytes for columnar records here, to avoid
+            // dual-counting the same records. (This means that our goodput % is much lower
+            // during the migration, which is an accurate reflection of reality.)
+            BlobTraceUpdates::Both(c, _) => c.goodbytes(),
+            BlobTraceUpdates::Structured {
+                key_values,
+                timestamps,
+                diffs,
+            } => {
+                key_values.goodbytes()
+                    + timestamps.values().to_byte_slice().len()
+                    + diffs.values().to_byte_slice().len()
+            }
+        }
+    }
+
+    /// Return the [`ColumnarRecordsStructuredExt`] of the blob.
+    pub fn get_or_make_codec<K: Codec, V: Codec>(
+        &mut self,
+        key_schema: &K::Schema,
+        val_schema: &V::Schema,
+    ) -> &ColumnarRecords {
+        loop {
+            match self {
+                BlobTraceUpdates::Row(records) => break records,
+                BlobTraceUpdates::Both(records, _) => break records,
+                BlobTraceUpdates::Structured {
+                    key_values,
+                    timestamps,
+                    diffs,
+                } => {
+                    let key =
+                        schema2_to_codec::<K>(key_schema, &*key_values.key).expect("valid keys");
+                    let val =
+                        schema2_to_codec::<V>(val_schema, &*key_values.val).expect("valid values");
+                    let records = ColumnarRecords::new(key, val, timestamps.clone(), diffs.clone());
+                    *self = BlobTraceUpdates::Both(records, key_values.clone());
+                }
+            }
+        }
```

Comment:
The `loop` is a bit scary to me because we could set hung, what do you think about this?

```suggestion
        match self {
            BlobTraceUpdates::Row(records) => records,
            BlobTraceUpdates::Both(records, _) => records,
            BlobTraceUpdates::Structured {
                key_values,
                timestamps,
                diffs,
            } => {
                let key =
                    schema2_to_codec::<K>(key_schema, &*key_values.key).expect("valid keys");
                let val =
                    schema2_to_codec::<V>(val_schema, &*key_values.val).expect("valid values");
                let records = ColumnarRecords::new(key, val, timestamps.clone(), diffs.clone());
                *self = BlobTraceUpdates::Both(records, key_values.clone());
                let BlobTraceUpdates::Both(records, _) = &self else {
                    unreachable!("programming error")
                }
                
                records
            }
        }
```